### PR TITLE
fix issue 210: custom values formatting schema

### PR DESCRIPTION
**Pull request by goodwin64**
Resolves https://github.com/sindresorhus/query-string/issues/210

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -250,7 +279,9 @@ function parse(input, options) {
 		return ret;
 	}
 
-	return (options.sort === true ? Object.keys(ret).sort() : Object.keys(ret).sort(options.sort)).reduce((result, key) => {
```
Comment (goodwin64) [inline index.js]: Too much logic in the `return` statement makes it hard to understand what's actually returned from the function.

File: index.d.ts [hunk 2]
```diff
@@ -1,3 +1,5 @@
+export type CustomValueParser = (value: string) => any;
```
Comment (sindresorhus) [inline index.d.ts]:
```suggestion
export type CustomValueParser = (value: string) => unknown;
```

File: index.js [hunk 3]
```diff
@@ -64,62 +64,90 @@ function encoderForArrayFormat(options) {
...
+	}
+
+	if (typedSchemaFormatter === 'number') {
+		return Number;
+	}
+
+	if (typeof typedSchemaFormatter === 'function') {
+		return typedSchemaFormatter;
+	}
+
```
Comment (sindresorhus) [inline index.js]: It should throw if the `typedSchemaFormatter` type is not a supported type.

File: test/parse.js [hunk 4]
```diff
@@ -310,3 +310,19 @@ test('query strings having comma encoded and format option as `comma`', t => {
...
+			baz: val => parseFloat(val) * 2
+		}
+	});
+	const expected = {
+		foo: '1',
+		bar: 1,
+		baz: 3
+	};
+	t.deepEqual(actual, expected);
+});
```
Comment (sindresorhus) [inline test/parse.js]: This feature will need a lot more tests.

File: test/parse.js [hunk 5]
```diff
@@ -310,3 +310,19 @@ test('query strings having comma encoded and format option as `comma`', t => {
 		]
 	});
 });
+
+test('custom parameters parsing', t => {
+	const actual = queryString.parse('foo=1&bar=1&baz=1.5', {
+		types: {
+			foo: 'string',
+			bar: 'number',
+			baz: val => parseFloat(val) * 2
```
Comment (sindresorhus) [inline test/parse.js]:
```suggestion
			baz: value => Number.parseFloat(value) * 2
```

File: index.d.ts [hunk 6]
```diff
@@ -121,6 +123,30 @@ export interface ParseOptions {
...
+			foo: 'string',
+			bar: 'number',
+			baz: val => parseFloat(val) * 2
+		}
+	});
+	//=> {foo: '1', bar: 1, baz: 3}
+	```
+	*/
+	readonly types?: {
+		[key: string]: 'string' | 'number' | CustomValueParser;
```
Comment (sindresorhus) [inline index.d.ts]:
```suggestion
		[type: string]: 'string' | 'number' | CustomValueParser;
```

File: readme.md [hunk 7]
```diff
@@ -181,6 +181,27 @@ queryString.parse('foo=true', {parseBooleans: true});
...
+Type: `object`\
+Default: `{}`
+
+```js
+const queryString = require('query-string');
+
+queryString.parse('foo=1&bar=1&baz=1.5', {
+    types: {
+        foo: 'string',
+        bar: 'number',
```
Comment (sindresorhus) [inline readme.md]: What about an array of string or array of numbers? We should probably use the notation `string[]` and `number[]` for those.

Comment (goodwin64) [reply readme.md]: Agree

**Comment by goodwin64** [comment]
Sorry, I've just realized unfortunately I have no spare time to finish the PR. If someone is willing to finish - it'd cool to finally deliver this functionality.

**Pull request closed without merging**